src/pybind/mgr/cephadm/module.py

```diff
@@ -2785,7 +2785,18 @@ def upgrade_status(self):
         return trivial_result(r)
 
     def upgrade_start(self, image, version):
+        if self.mode != 'root':
+            raise OrchestratorError('upgrade is not supported in %s mode' % (
+                self.mode))
         if version:
+            try:
+                (major, minor, patch) = version.split('.')
+                assert int(minor) >= 0
+                assert int(patch) >= 0
+            except:
+                raise OrchestratorError('version must be in the form X.Y.Z (e.g., 15.2.3)')
+            if int(major) < 15 or (int(major) == 15 and int(minor) < 2):
+                raise OrchestratorError('cephadm only supports octopus (15.2.0) or later')
```

Comment:
how will this work for the monitoring components (node-exporter, prom, grafana)?

Reply:
it wont.  this only supports upgrading ceph (currently)